core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,416 @@
+            future = cache.putAsync(key, mergedEntry.getValue(), mergedEntry.getMetadata());
+         }
+         future.whenComplete((responseMap, exception) -> {
+            phaser.arriveAndDeregister();
+            if (exception != null)
+               log.exceptionDuringConflictResolution(key, exception);
+         });
+      });
+   }
+
+   @Override
+   public boolean isStateTransferInProgress() {
+      return stateConsumer.isStateTransferInProgress();
+   }
+
+   private class VersionRequest {
+      final K key;
+      final boolean postpone;
+      final CompletableFuture<Map<Address, InternalCacheValue<V>>> completableFuture = new CompletableFuture<>();
+      volatile CompletableFuture<Map<Address, Response>> rpcFuture;
+      volatile Collection<Address> keyOwners;
+
+      VersionRequest(K key, boolean postpone) {
+         this.key = key;
+         this.postpone = postpone;
+
+         if (trace) log.tracef("Creating %s", this);
+
+         if (postpone) {
+            retryQueue.add(this);
+         } else {
+            start();
+         }
+      }
+
+      void cancelRequestIfOutdated() {
+         Collection<Address> latestOwners = installedTopology.getWriteOwners(key);
+         if (rpcFuture != null && !completableFuture.isDone() && !keyOwners.equals(latestOwners)) {
+            rpcFuture = null;
+            keyOwners.clear();
+            if (rpcFuture.cancel(false)) {
+               retryQueue.add(this);
+
+               if (trace) log.tracef("Cancelling %s for nodes %s. New write owners %s", this, keyOwners, latestOwners);
+            }
+         }
+      }
+
+      void start() {
+         // TODO figure out if we can just use the CacheTopology.getPendingCh with segment id for a key instead of creating a mock LocalizedCacheTopology
+         // cacheTopology.getSegment(key)
+         //  cacheTopology.getPendingCH().locateOwnersForSegment();
+         // It should be fine, as keyPartitoner should not change during runtime, so we're basically using a non stored version of what is
+         // in a localized topology
+         keyOwners = installedTopology.getWriteOwners(key);
+
+         if (trace) log.tracef("Attempting %s from owners %s", this, keyOwners);
+
+         final Map<Address, InternalCacheValue<V>> versionsMap = new HashMap<>();
+         if (keyOwners.contains(localAddress)) {
+            GetCacheEntryCommand cmd = commandsFactory.buildGetCacheEntryCommand(key, localFlags);
+            InvocationContext ctx = invocationContextFactory.createNonTxInvocationContext();
+            InternalCacheEntry<K, V> internalCacheEntry = (InternalCacheEntry<K, V>) interceptorChain.invoke(ctx, cmd);
+            InternalCacheValue<V> icv = internalCacheEntry == null ? null : internalCacheEntry.toInternalCacheValue();
+            versionsMap.put(localAddress, icv);
+         }
+
+         ClusteredGetCommand cmd = commandsFactory.buildClusteredGetCommand(key, FlagBitSets.SKIP_OWNERSHIP_CHECK);
```

Comment:
please use `cmd.setTopologyId(installedTopology.getTopologyId())` to mark which topology you use for sending